iroha/src/sumeragi.rs

```diff
@@ -1,14 +1,216 @@
+    Created(Block),
+    /// Is sent by validating peers to proxy tail and observing peers when they have signed this block.
+    Signed(Block),
+    /// Is sent by proxy tail to validating peers and to leader, when the block is committed.
+    Committed(Block),
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Role {
+    Leader,
+    ValidatingPeer,
+    ObservingPeer,
+    ProxyTail,
+}
 
 impl Sumeragi {
-    pub fn new() -> Self {
-        Sumeragi {}
+    pub fn new(
+        peers: &[PeerId],
+        prev_block_hash: Option<Hash>,
+        max_faults: usize,
+    ) -> Result<Self, String> {
+        let min_peers = 3 * max_faults + 1;
+        if peers.len() >= min_peers {
+            let mut peers = peers.to_vec();
+            Self::sort_peers(&mut peers, prev_block_hash);
+            Ok(Self {
+                sorted_peers: peers,
+                max_faults,
+                //TODO: generate peer's public key, save on shutdown and load on start
+                peer_id: PeerId {
+                    address: "127.0.0.1:7878".to_string(),
+                    public_key: [0u8; 32],
+                },
+                pending_block: None,
+            })
+        } else {
+            Err(format!("Not enough peers to be Byzantine fault tolerant. Expected a least {} peers, got {}", 3 * max_faults + 1, peers.len()))
+        }
+    }
+
+    pub fn next_round(&mut self, prev_block_hash: Hash) {
+        Self::sort_peers(&mut self.sorted_peers, Some(prev_block_hash));
+        self.pending_block = None;
+    }
+
+    pub fn peers_set_a(&self) -> &[PeerId] {
+        let n_a_peers = 2 * self.max_faults + 1;
+        &self.sorted_peers[..n_a_peers]
     }
 
-    pub async fn sign(
+    pub fn peers_set_b(&self) -> &[PeerId] {
+        &self.sorted_peers[(2 * self.max_faults + 1)..]
+    }
+
+    pub fn leader(&self) -> &PeerId {
+        self.peers_set_a()
+            .first()
+            .expect("Failed to get first peer.")
+    }
+
+    pub fn proxy_tail(&self) -> &PeerId {
+        self.peers_set_a().last().expect("Failed to get last peer.")
+    }
+
+    pub fn validating_peers(&self) -> &[PeerId] {
+        &self.peers_set_a()[1..(self.peers_set_a().len() - 1)]
+    }
+
+    pub fn role(&self) -> Role {
+        if *self.leader() == self.peer_id {
+            Role::Leader
+        } else if *self.proxy_tail() == self.peer_id {
+            Role::ProxyTail
+        } else if self.validating_peers().contains(&self.peer_id) {
+            Role::ValidatingPeer
+        } else {
+            Role::ObservingPeer
+        }
+    }
+
+    pub fn validate_access(&self, allowed_roles: &[Role]) -> Result<(), String> {
+        if allowed_roles.contains(&self.role()) {
+            Ok(())
+        } else {
+            Err(format!(
+                "Peer needs to be one of {:?} for this operation.",
+                allowed_roles
+            ))
+        }
+    }
+
+    fn sort_peers(peers: &mut Vec<PeerId>, block_hash: Option<Hash>) {
```

Comment:
It might be more efficient to seed a random number generator with the previous block hash and then use that to determine the order.